Feature: Warn if destination directory is formatted as ntfs

fallingrock opened the pull request:
Warn user if destination directory is formatted as ntfs to avoid "Unsupported operation" errors.

Fix #1854

fallingrock committed:
Feature: Warn if destination directory is formatted as ntfs

buhtz commented on qt/settingsdialog.py:
```diff
@@ -15,6 +15,7 @@
 import copy
 import re
 import getpass
+import psutil
```
I wouldn't like to add a new dependency for that use case.

Why not using the shell tools?

    >>> r = subprocess.run(['df', '--output=fstype', '.'], stdout=subprocess.PIPE, text=True)
    >>> r.stdout.splitlines()[1]
    'ext4'

psutils also only read from the /proc filesystem.

**EDIT**:
BIT still hast this "feature". See https://github.com/bit-team/backintime/blob/8c7afef81ae79be491a848400261eabdc3e90c8c/common/tools.py#L1450

fallingrock replied:
Wasn't aware of the filesystem method ... I'll check it out.

fallingrock commented:
Hmmm ... build failed because psutil wasn't available.

Can I add psutil to the travis.yml file to fix this?

buhtz reviewed (changes requested):
I wonder if this will help in all cases.

The file system depends on the way how the ntfs volume is mounted.
How was the volume mounted in case of the issue you reported? Would it catch that case?

buhtz commented on CHANGES:
```diff
@@ -1,6 +1,7 @@
 Back In Time
 
 Version 1.5.3-dev (development of upcoming release)
+* Feature: Warn if destination directory is formatted as NTFS
```
```suggestion
* Feature: Warn if destination directory is formatted as NTFS (#1854) (David Gibbs @fallingrock)
```

fallingrock committed:
Update CHANGES

Co-authored-by: buhtz <c.buhtz@posteo.jp>

fallingrock committed:
Use existing filesystem determination method and remove psutil dependency.

buhtz commented on qt/settingsdialog.py:
```diff
@@ -1947,6 +1947,14 @@ def btnSnapshotsPathClicked(self):
                 self.config.removeProfileKey('snapshots.path.uuid')
 
+            if tools.filesystem(path).startswith('ntfs'):
+                text = _('Warning: {0} is formatted as NTFS which has known incompatabilities '
+                         'with unix style file systems.\n\n'
+                         'Are you sure you want to use it as your backup destination?')
+                question = text.format(path)
```
Imagine our translators using the web-frontend at translators.codeberg.org. Some of them have no coding background. They also need more context. Because of that it is important to give placeholders good names. The details are [hidden in our docs](https://github.com/bit-team/backintime/blob/dev/common/doc-dev/2_localization.md#instructions-for-the-translation-process).

```suggestion
                text = _('Warning: {path} is formatted as NTFS which has known incompatibilities '
                         'with Unix style file systems.'
                         'Are you sure you want to use it as your backup destination?')
                question = text.format(path=path)
```

I also removed the two `\n` because of our translators. I am not sure if, so please checkout if the dialog box does wrap the text by itself. And while you are checking it please create a screenshot of message box and attach it to the PR. If the text is not wrapped, please simply make two translatable strings like this:

    text = _('foo') + '\n\n' + _('bar')

Beside the text, the translators will also get a screenshot of the message box. So they will understand.

It is damn hard to get translators and keep them motivated. So I try to make it as easy as possible for them. 😄

buhtz commented:
The message seems a bit unspecific for some users. Are there cases where NTFS volumes do work? So why are the users free to use NTFS with BIT? Shouldn't we just deny such destination volumes?

fallingrock commented:
> The message seems a bit unspecific for some users. Are there cases where NTFS volumes do work? So why are the users free to use NTFS with BIT? Shouldn't we just deny such destination volumes?

I think it depends on if the user has filenames containing certain characters.

The following reserved characters:   
 \> (less than)
 \< (greater than)   
 : (colon)   
 " (double quote)   
 / (forward slash)   
 \ (backslash)   
 | (vertical bar or pipe)   
 ? (question mark)   
 \* (asterisk)

fallingrock commented:
Screenshot of new error messages
![image](https://github.com/user-attachments/assets/d4d44566-b0aa-4aea-8369-15360c9e505f)
![image](https://github.com/user-attachments/assets/f4f0e465-7d3d-49ea-b0b5-2d1667c1fccc)

fallingrock committed:
Minor change to warning dialog text.

fallingrock committed:
Added additional ntfs check in setSnapshotsPath method and made messages
consistent.

buhtz commented:
> I think it depends on if the user has filenames containing certain characters.
> The following reserved characters: ...

Interesting. In this case my alternative suggested approach just doing an rsync test run and check after that if hardlinks appear won't have caught such a situation. Your approach seems better.

Would you like to add an entry in FAQ.md about that NTFS details? We later can move that FAQ entry somewhere else into the docu.

buhtz commented on common/config.py:
```diff
@@ -494,6 +494,12 @@ def setSnapshotsPath(self, value, profile_id = None, mode = None):
 
+        elif fs.startswith('ntfs'):
+            self.notifyError(
+                _('Destination filesystem for {path} is formatted with NTFS '
+                  'which has known incompatibilities with Unix style file systems.')
```
I am not skilled with English words. I just asked the "stochastic parrot" and modified the string a bit. So feel free to modify it either.

```suggestion
                _('The destination filesystem for {path} is formatted with NTFS, '
                  'which has known incompatibilities with Unix-style filesystems.')
```

fallingrock replied:
I'm no english major either (GIBBeriSh is my family language :) ) ... I was just matching the existing message pattern in config.py.

buhtz commented on qt/settingsdialog.py:
```diff
@@ -1947,6 +1947,15 @@ def btnSnapshotsPathClicked(self):
 
+            fs = tools.filesystem(path)
+            if fs.startswith('ntfs'):
+                text = _('Warning: Destination filesystem for {path} is formatted with NTFS '
+                         'which has known incompatibilities with Unix style file systems.\n'
+                         'Are you sure you want to use it as your backup destination?')
```
Beside avoiding escape characters like `\n` in the translation string I also try to avoid referencing user in person via "you".

```suggestion
                text = '\n'.join([
                       _('Warning: The destination filesystem for {path} is formatted with NTFS, '
                         'which has known incompatibilities with Unix-style filesystems.'),                             
                       _('Is this the backup destination to be used?')
                })
```

buhtz reviewed (changes requested):
The PR looks good. Thank you very much for your efforts. A FAQ entry would be the cherry on top.

I would like to wait for review from the other team mates: emtiu and/or aryoda. They are not quit available currently so it can take some time.

fallingrock committed:
Update qt/settingsdialog.py

Co-authored-by: buhtz <c.buhtz@posteo.jp>

fallingrock committed:
Added FAQ entry
Moved warning message text to common location
Corrected syntax error on text join

fallingrock committed:
Merge branch 'bit-team:dev' into ntfs-check

buhtz committed:
Merge branch 'dev' into ntfs-check
